mslice/models/workspacemanager/workspace_algorithms.py

```diff
@@ -107,6 +107,7 @@ def _get_property_from_history(name, history):
 
 
 def get_q_limits(theta, emax, efix):
+    # calculate |Q| limits in Angstrom^-1 from theta (scattering angle) in radians
     qmin, qmax, qstep = tuple(np.sqrt(E2q * 2 * efix * (1 - np.cos(theta)) * meV2J) / m2A)
     qmax = np.sqrt(E2q * (2 * efix + emax - 2 * np.sqrt(efix * (efix + emax)) * np.cos(theta[1])) * meV2J) / m2A
     return qmin, qmax, qstep
```

Comment:
@MatthewBowles Thanks for looking into this. The Q-limits bug can be fixed by changing the code here:

```
--- a/mslice/models/workspacemanager/workspace_algorithms.py
+++ b/mslice/models/workspacemanager/workspace_algorithms.py
@@ -68,9 +68,8 @@ def _processLoadedWSLimits(workspace):
 def process_limits(ws):
     en = ws.raw_ws.getAxis(0).extractValues()
     theta = _get_theta_for_limits(ws)
-    # Use minimum energy (Direct geometry) or maximum energy (Indirect) to get qmax
-    emax = -np.min(en) if (str(ws.e_mode == 'Direct')) else np.max(en)
-    qmin, qmax, qstep = get_q_limits(theta, emax, ws.e_fixed)
+    # If Direct geometry use +e_fixed, if indirect geometry, use -e_fixed
+    qmin, qmax, qstep = get_q_limits(theta, en, ws.e_fixed if (str(ws.e_mode == 'Direct')) else -ws.e_fixed)
     set_limits(ws, qmin, qmax, qstep, theta, np.min(en), np.max(en), np.mean(np.diff(en)))
 
 
@@ -106,9 +105,12 @@ def _get_property_from_history(name, history):
     return None
 
 
-def get_q_limits(theta, emax, efix):
-    qmin, qmax, qstep = tuple(np.sqrt(E2q * 2 * efix * (1 - np.cos(theta)) * meV2J) / m2A)
-    qmax = np.sqrt(E2q * (2 * efix + emax - 2 * np.sqrt(efix * (efix + emax)) * np.cos(theta[1])) * meV2J) / m2A
+def get_q_limits(theta, en, efix):
+    Ei = np.abs(efix)
+    qlines = [np.sqrt(E2q * (2 * Ei - en - 2 * np.sqrt(Ei * (Ei - en)) * np.cos(tth)) * meV2J) / 1e10 for tth in theta[:2]]
+    qmin = np.min(qlines[0 if (efix > 0) else 1])
+    qmax = np.max(qlines[1 if (efix > 0) else 0])
+    qstep = np.sqrt(E2q * 2 * efix * (1 - np.cos(theta[2])) * meV2J) / m2A
     return qmin, qmax, qstep
```

Reply:
Thanks!